project/02-Modularity/Handout.md

```diff
@@ -0,0 +1,160 @@
+
+Now that you know how your system should work, you can write interfaces and data classes that _describe_ this behaviour. Don't think too much about the actual implementation, just write an interface that exposes the behaviour you need.
+
+In theory, you will have three interfaces (the names are indicative):
+ - `LocationService` that defines a method to get the current position of the device
+ - `GeoCodingService` that defines a method to convert a position to an address (the reverse can also be useful if you want to display it to the user)
+ - `WeatherService` that defines a method to get the weather at a given location.
+
+You will also need a few classes to store the resulting forecast from the server. Think about _what values your client will need_ (while still making sure you can provide them - have a look at the [OneCall API documentation](https://openweathermap.org/api/one-call-api)). A `Location` class will also be needed to store the location - both when retrieving it in and when sending it to the weather API. Android already provides one, but you don't want to use it here, as it would not cleanly allow you to separate concerns and use an other provider in the future with the same interfaces.
+
+Regarding the design of the classes storing the forecast, here are a few advices. The API returns data in Json (a sort of key-value object notation), containing in particular the current weather and the forecast for the following days. See which fields interest you and design the interfaces and classes you need.
+
+> :information_source: When you design a service that _waits_ on something (for example: an external API), you usually don't want to block the caller, but rather use asynchronous structures. We will introduce them later in the course, so **we don't ask you to use them for now**. **However, remember that -outside of this exercise- calling a web service synchronously from the main thread is in most of the cases a very bad practice.** 
+
+## Implement the interfaces
+
+Now that you wrote the interfaces for your modules, you need to implement them. For each interface, create a class that implements the methods you defined. Here are a few hints and tips on what to do precisely. Feel free to ask on Piazza if you need suggestions!
+
+> :information_source: You may notice while writing the implementation that the interface misses something: a method parameter, a list of thrown exceptions... This is not that big of a deal: sometimes, when writing the actual code, you come up with something you didn't think you'd need. Updating the interface at this time to provide what you need for that is not an issue. **However, please go through the same reasoning every time you update the interface as when you wrote it: is it specific to my implementation? Does it give too much details about the implementation? Do I really need it? ...**. 
```

Comment:
Updating the interface at this time to provide what you need for that is not an issue
->
Updating the interface at this stage to provide what you need is not an issue